File: java/src/main/java/com/authdog/resources/EmailProvidersResource.java
package com.authdog.resources;

import com.authdog.AuthdogClient;
import com.authdog.exceptions.ApiException;
import com.authdog.exceptions.AuthenticationException;
import com.fasterxml.jackson.databind.JsonNode;

/**
 * Environment email provider operations.
 */
public final class EmailProvidersResource {
    /**
     * Authdog client.
     */
    private final AuthdogClient client;

    /**
     * Create an email providers helper.
     * @param clientParam Authdog client
     */
    public EmailProvidersResource(final AuthdogClient clientParam) {
        this.client = clientParam;
    }

    /**
     * List email providers.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @return providers envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode list(final String tenantId,
                         final String environmentId)
            throws AuthenticationException, ApiException {
        return client.request("GET",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/email-providers",
                null, null, JsonNode.class);
    }

    /**
     * Create or update an email provider.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return provider envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode save(final String tenantId,
                         final String environmentId,
                         final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/email-providers",
                body, null, JsonNode.class);
    }

    /**
     * Test an email provider.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param body request body
     * @return test envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode test(final String tenantId,
                         final String environmentId,
                         final Object body)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/email-providers/test",
                body, null, JsonNode.class);
    }

    /**
     * Delete an email provider.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param provider provider name
     * @return delete envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode delete(final String tenantId,
                           final String environmentId,
                           final String provider)
            throws AuthenticationException, ApiException {
        return client.request("DELETE",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/email-providers/" + provider,
                null, null, JsonNode.class);
    }

    /**
     * Activate an email provider.
     * @param tenantId tenant ID
     * @param environmentId environment ID
     * @param provider provider name
     * @return activate envelope
     * @throws AuthenticationException when unauthorized
     * @throws ApiException when the request fails
     */
    public JsonNode activate(final String tenantId,
                             final String environmentId,
                             final String provider)
            throws AuthenticationException, ApiException {
        return client.request("POST",
                EnvPaths.prefix(tenantId, environmentId)
                        + "/email-providers/" + provider
                        + "/activate",
                null, null, JsonNode.class);
    }
}

File: csharp/Resources/EmailProvidersResource.cs
using System.Net.Http;
using System.Threading.Tasks;
using Newtonsoft.Json.Linq;

namespace Authdog
{
    /// <summary>
    /// Environment email-provider Wave 3 operations
    /// </summary>
    public class EmailProvidersResource
    {
        private readonly AuthdogClient _client;

        public EmailProvidersResource(AuthdogClient client)
        {
            _client = client;
        }

        public Task<JObject> ListAsync(string tenantId, string environmentId) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Get,
                $"{AuthdogClient.Env(tenantId, environmentId)}/email-providers");

        public JObject List(string tenantId, string environmentId) =>
            ListAsync(tenantId, environmentId).GetAwaiter().GetResult();

        public Task<JObject> SaveAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/email-providers",
                body);

        public JObject Save(string tenantId, string environmentId, object body) =>
            SaveAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> TestAsync(string tenantId, string environmentId, object body) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/email-providers/test",
                body);

        public JObject Test(string tenantId, string environmentId, object body) =>
            TestAsync(tenantId, environmentId, body).GetAwaiter().GetResult();

        public Task<JObject> DeleteAsync(string tenantId, string environmentId, string provider) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Delete,
                $"{AuthdogClient.Env(tenantId, environmentId)}/email-providers/{provider}");

        public JObject Delete(string tenantId, string environmentId, string provider) =>
            DeleteAsync(tenantId, environmentId, provider).GetAwaiter().GetResult();

        public Task<JObject> ActivateAsync(string tenantId, string environmentId, string provider) =>
            _client.RequestAsync<JObject>(
                HttpMethod.Post,
                $"{AuthdogClient.Env(tenantId, environmentId)}/email-providers/{provider}/activate");

        public JObject Activate(string tenantId, string environmentId, string provider) =>
            ActivateAsync(tenantId, environmentId, provider).GetAwaiter().GetResult();
    }
}
